frontend/src/hooks/useAuth.js

```diff
@@ -0,0 +1,171 @@
+ * @property {Function} login - Function to log in a user with provided credentials.
+ * @property {Function} logout - Function to log out the user.
+ * @property {string|null} loginError - An error message if login fails, or null if there's no
+ * error.
+ * @property {Function} setLoginError - Function to set the login error message.
+ * @property {Function} setRedirect - Function to set the redirect URL after login.
+ * @property {Function} getAccessToken - Function to retrieve and refresh the access token.
+ */
+export default function useAuth() {
+  const router = useRouter();
+  const [user, setUser] = useState(null);
+  const [accessToken, setAccessToken] = useState(null);
+  const [loginError, setLoginError] = useState(null);
+  const [isLoading, setIsLoading] = useState(true);
+  const [redirect, setRedirect] = useState('/');
+
+  const addRefreshToken = (refreshToken) => {
+    Cookies.set('refresh', refreshToken, { secure: true });
+  };
+
+  const removeRefreshToken = () => {
+    Cookies.remove('refresh');
+  };
+
+  const getRefreshToken = () => Cookies.get('refresh');
+
+  const removeAccessToken = () => {
+    setAccessToken(null);
+  };
+
+  const addAccessToken = (token) => {
+    setAccessToken(token);
+  };
+
+  const login = async (userData) => {
+    try {
+      const response = await axios.post(LOGIN_SVC_URI, userData);
+      const { access, refresh } = response.data;
+      const { user_role } = decode(access);
+
+      const newUser = {
+        username: userData.username,
+        role: user_role,
+      };
+
+      setUser(newUser);
+      addAccessToken(access);
+      addRefreshToken(refresh);
+      router.push(redirect);
+    } catch (err) {
+      if (err.response && err.response.status === 401) {
+        console.error('Unauthorized: ', err.response.data);
+        setLoginError('The username or password you entered is incorrect');
+      } else {
+        console.error('An error occurred: ', err);
+        setLoginError('An error occurred. Please try again later.');
+      }
+    }
+  };
+
+  const logout = () => {
+    setUser(null);
+    removeAccessToken();
+    removeRefreshToken();
+    // TODO: redirect to '/', which redirects to /login if not auth
+    router.push('/login');
+  };
+
+  const refreshAccessToken = useCallback(async () => {
+    const refreshToken = getRefreshToken();
+    let newAccessToken;
+    if (!refreshToken) {
+      return;
+    }
+
+    const body = { refresh: refreshToken };
+
+    try {
+      console.log('refreshing token...');
+      const response = await axios.post(REFRESH_TOKEN_SVC_URI, body);
+      const { access, refresh } = response.data;
+      newAccessToken = access;
+      addAccessToken(access);
+      addRefreshToken(refresh);
+
+      // KIV: not advisable to store user data in tokens, best to make another req to get user info
+      const { username, user_role } = decode(refresh);
+
```

Comment:
I am thinking of removing username and user_role information from the refresh token, but not for the access token. What do you think?